README
fmtools: simple V4L2 radio card programs

Maintained by Ben Pfaff <blp@cs.stanford.edu>
Originally by Russell Kroll <rkroll@exploits.org>

Web: http://benpfaff.org/fmtools
Git: git://repo.or.cz/fmtools.git

License
=======

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

Package information
===================

This is a pair of hopefully useful control programs for Video for
Linux 2 (V4L2) radio card drivers.  The focus is on control, so you
may find these programs a bit unfriendly.  Users are encouraged to
investigate the source and create wrappers or new programs based on
this design.

 fm      - a simple tuner
 fmscan  - a simple band scanner

fm
==

This is a simple program that will send commands to your V4L2 radio
cards.  It was written mostly so I could have something to send
commands at the drivers being created or debugged here.

The usual mode of operation is to tell the radio to come on to a given
freqency.  In these examples, 94.3 will be used since that's a station
that happens to get tuned here frequently.

To turn the radio on to that frequency at the default volume, you'd do
"fm 94.3" and call it done.  If you wanted to crank it up to full
volume, another argument would be needed and the command becomes "fm
94.3 100".  Volumes range from 0 (off) to 100 (full intensity).

There is a -q option that will suppress messages about tuning, volume
changes, and so forth.  Keeping your tuner quiet may not seem like
such a useful thing at first, but it can be very handy inside an IRC
script.  Having a program scribble on your channel window when you do
/KILO or similar is not enjoyable.

Besides direct tuning, there are also "on" and "off" commands.  They
will turn the card on and off (mute) as you may have guessed.

Finally, there is volume control.  To go up a notch, use +.  To go
down a notch, use -.  Simple.

fm configuration
----------------

You can create a file called .fmrc in your home directory that
contains values for the default volume and default volume stepping.
This way, you can have your radio card always come on at a certain
volume when you don't explicitly list it on the command line.  The
volume stepping controls how much the volume changes during a - or +
operation.

This file is not required for operation of the program, and it will
not even give so much as a peep if it doesn't find it.  The defaults
are used when you don't specify values.  The values are 12.5% for
default volume and 10% for default volume stepping.

Here's what a real .fmrc might look like ...

VOL 50
INCR 10

Here we say the default volume is 50%.  Then we say the increment
value for volume changes with + and - is 10%.  These values were
obtained by playing around with the numbers until things behaved the
way I wanted.  Be aware that some cards only have a couple of volume
steps, so you may not notice a change right away.

The Cadet hardware, for example, has two settings: off and *VERY
LOUD*.  Any volume setting will turn it on.  The Aztech card has 4
settings (off, low, medium, high) and you'll notice it change when you
cross the boundaries.  The Aimslab RadioReveal card is actually based
on analog volume control, with 10 steps emulated within that range.

There are some other options available:

-d - select device.  If you have more than one V4L2 radio card in your
     system for some reason, use -d <device> to select the right one.
     This is probably only useful if you like writing drivers for
     these things.

-o - override card frequency range - Some radio card drivers don't
     actively enforce the frequencies that you can tune to.  Use this
     switch and fm will send any frequency you want to the driver.
     This alone will not make an unmodified card receive frequencies
     from outside the design range.  You have to change the tuner
     hardware in order to enjoy such broadcasts.

-t - select tuner - Certain cards have multiple tuners - usually used
     for different bands.  The ADS Cadet driver is known to support
     this, specifically using 0 for FM and 1 for AM.  To tune AM
     frequencies, remember that fm expects MHz style input, and do the
     following:

     For 1200 kHz, enter 'fm -t 1 1.2', since 1.2 MHz == 1200 kHz.

     Yes, tuning am frequencies with 'fm' seems backwards, but that's
     life.  Crafty hackers may want to add some code to make the
     program check argv[0] and behave differently.

fmscan
======

This simple little program will command your radio card through the
radio band and show which ones have a accumulated signal strength of
50% or higher.  This process can take awhile, and can vary greatly
depending on the radio card in use.

By default, the range scanned is 87.9-107.9 MHz in .2 MHz steps, since
that's the standard band here in the USA.  Users in other regions
should set the appropriate information for best results.

V4L2 /dev entries
================

By default, these programs use /dev/radio0 to access the hardware.  This
can be changed with -d <device> if you have more than one.   If you have
been using v4l radio cards for awhile, you may already have a /dev/radio.
That is now "legacy", and should be symlinked to /dev/radio0.  To create
the proper device entry, either do "make devices" as root, or create it
by hand with mknod (c 81 64).

The proper device listing looks something like this in 'ls -la' ...

lrwxrwxrwx   1 root     root           11 Jan 20 03:19 /dev/radio -> /dev/radio0
crw-r--r--   1 root     root      81,  64 Jan 20 03:19 /dev/radio0

Other cards follow the same pattern - radio1 would be (81, 65), and so on.

The owner and permission data should be set to values that agree with
your system's personality.  If it's mostly a solitary system with few or
no users, the above settings will be fine.

However, if you have other people running around on your system,  consider 
making the device part of a "radio" group or maybe even "console" if you
use such a thing.  That will keep random individuals from doing odd things
to your radio like changing it to a classical station while you're 
listening to some death metal (or vice versa).  You have been warned.


NEWS
2.0.3:

Sat May  1 11:51:16 PDT 2010 / Ben Pfaff <blp@cs.stanford.edu>

- Debian packaging update.

2.0.2:

Sat Mar 27 10:53:43 PDT 2010 / Ben Pfaff <blp@cs.stanford.edu>

- Debian packaging update.

2.0.1:

Sat Jan 9 13:28:53 PST 2010 / Ben Pfaff <blp@cs.stanford.edu>

- Fix bugs that prevented fm and fmscan from working outside "test"
  mode (used by the testsuite.

- Better support radios that do not have volume control.

- Make fm report when the radio is muted and tell the user how to unmute it.

- Packaging improvements.

2.0:

Sat Dec 12 21:18:52 PST 2009 / Ben Pfaff <blp@cs.stanford.edu>

- Rewrite to use video4linux2 API.

- Add "-q" (quiet) and "-T" (tuner) options to fmscan.

- Improve fmscan frequency calculations.  It should now be better at
  accurately finding radio stations.

- Add Autotest test suite and, in the process, convert the packaging
  to use Autoconf and Automake.

- Some code and some documentation used a range of 0-65535 for volume,
  other code and other documentation used a range of 0-100.  This has
  been consistently changed to 0-100.

1.0.2:

Thu Nov 23 1:41:00 PDT 2006 / Ben Pfaff <blp@cs.stanford.edu>

- Fix version number reported by fm and fmscan programs, as well as
  the version number included in the manpages.

  Bug report by "Petr Slansky" <slansky@usa.net>.

- No functional changes.

1.0.1:

Sat Oct 7 1:05:00 PDT 2006 / Ben Pfaff <blp@cs.stanford.edu>

- Make fm able to sleep for a user-specified time, or forever, after
  applying settings.  This makes fm usable with radio tuner drivers
  that turn off the radio after the fd used for tuning is closed.

  Idea and patch from Dave Ulrick <d-ulrick@insightbb.com>.

1.0:

Sat Sep 3 12:21:00 PDT 2004 / Ben Pfaff <blp@cs.stanford.edu>

- New maintainer.  New home page is http://benpfaff.org/fmtools

- Updated email addresses.

- Integrated Debian packaging.

- Made fmscan threshold configurable from command line.

- Use videodev.h header detached from kernel to avoid including kernel
  headers.

0.99.1:

Fri Aug 16 23:08:30 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm, fmscan: fix the FSF's address in the headers

Fri Aug 16 23:07:49 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - COPYING: put in the modern version of the GPL v2, with the correct
   address for the FSF.

Fri Aug 16 23:05:33 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - Makefile: reworked a bunch of stuff, and added such useful things
   like install targets.  Hopefully this will make packaging less of 
   a hack.

Fri Aug 16 22:55:41 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: do the VIDIOCGAUDIO first so there's always good data in the struct

Fri Aug 16 22:48:10 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm.1: incorporated the man page written by Ben Pfaff.

Fri Aug 16 22:47:01 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: set va.balance to 32768 rather than relying on whatever junk
   happens to be in the struct.   Reported by ahmet inan.

Fri Aug 16 22:43:47 CDT 2002 / Russell Kroll <rkroll@exploits.org>

 - fm: applied patch from Viktor Rosenfeld to fix the rounding so that
   channels like 99.95 MHz can be tuned.

0.99.0:

Sat Feb  3 22:41:12 CST 2001 - Russell Kroll <rkroll@exploits.org>

 - Final cleanups for 0.99.0.  This source has been sitting around for
   awhile, so it's probably time to push towards 1.0 and call it done.
   How much can you do to a pair of radio programs, anyway?

Thu Jan  4 17:29:49 CST 2001 - Russell Kroll <rkroll@exploits.org>

 - fmscan: ioctl/code cleanups, ala fm
 - fmscan: -d <dev> support
 - fm: -d <dev> support for multiple devices (/dev/radio1 and so on)
 - fm: ioctl results checked, other coding cleanups

0.2.5 and earlier:

Mon Apr 12 20:32:13 MDT 1999
 - -t option for selecting the tuner on multi-tuner boards (radio-cadet, etc)

Wed Feb  3 13:13:01 MST 1999
 - fmtools 0.2.4 released
 - 'devices' target added to Makefile
 - documentation updated for fmscan and /dev/radio0
 - programs now use /dev/radio0 instead of old(er) /dev/radio
 - fmscan added
 - small code cleanups
 - -o added to oneliner usage display

Sat Jan 16 23:55:15 MST 1999
 - fmtools 0.2.3 released
 - -o (override frequency range checks) introduced after some l-k discussion

Sun Jan 10 22:17:07 MST 1999
 - fmtools 0.2.2 released
 - overflows and negative values during volume changes fixed
 - exit codes are now sane
 - the division should now accurately reflect 100% volume

Thu Jul 30 08:39:50 MDT 1998
 - We now support fine tuning on cards that fly the VIDEO_TUNER_LOW flag.
   Thanks to Ben Pfaff for this addition.


fmlib.c
/* fmlib.c - simple V4L2 compatible tuner for radio cards

   Copyright (C) 2009 Ben Pfaff <blp@cs.stanford.edu>

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
   more details.

   You should have received a copy of the GNU General Public License along with
   this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "fmlib.h"
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

struct tuner_test {
        int freq;               /* In 1/16 MHz. */
        int volume;             /* Between 1000 and 2000. */
};

char *program_name;

static void query_control(const struct tuner *, uint32_t id,
                          struct v4l2_queryctrl *);
static int32_t get_control(const struct tuner *, uint32_t id);
static void set_control(const struct tuner *, uint32_t id, int32_t value);
static void query_tuner(const struct tuner *, struct v4l2_tuner *);

void
fatal(int error, const char *msg, ...)
{
        va_list args;

        fprintf(stderr, "%s: ", program_name);

        va_start(args, msg);
        vfprintf(stderr, msg, args);
        va_end(args);

        if (error)
                fprintf(stderr, ": %s", strerror(error));
        putc('\n', stderr);

        exit(EXIT_FAILURE);
}

static void *
xmalloc(size_t n)
{
        void *p = malloc(n ? n : 1);
        if (!p)
                fatal(0, "out of memory");
        return p;
}

void
tuner_open(struct tuner *tuner, const char *device, int index)
{
        memset(tuner, 0, sizeof *tuner);

        if (!device)
                device = "/dev/radio0";
        else if (!strcmp(device, "test") || !strncmp(device, "test ", 5)) {
                double volume;

                if (sscanf(device, "test %lf", &volume) != 1)
                        volume = 50;

                tuner->test = xmalloc(sizeof *tuner->test);
                tuner->test->freq = 90 * 16;
                tuner->test->volume = volume >= 0 ? volume * 10 + 1000.5 : 0;

                device = "/dev/null";
        }

	tuner->fd = open(device, O_RDONLY);
	if (tuner->fd < 0)
                fatal(errno, "Unable to open %s", device);
        tuner->index = index;

        query_control(tuner, V4L2_CID_AUDIO_VOLUME, &tuner->volume_ctrl);
        query_tuner(tuner, &tuner->tuner);
}

void
tuner_close(struct tuner *tuner)
{
        close(tuner->fd);
}

void
tuner_set_mute(struct tuner *tuner, bool mute)
{
        set_control(tuner, V4L2_CID_AUDIO_MUTE, mute);
}

bool
tuner_has_volume_control(const struct tuner *tuner)
{
        const struct v4l2_queryctrl *vqc = &tuner->volume_ctrl;
        return vqc->maximum > vqc->minimum;
}

double
tuner_get_volume(const struct tuner *tuner)
{
        if (tuner_has_volume_control(tuner)) {
                const struct v4l2_queryctrl *vqc = &tuner->volume_ctrl;
                int volume = get_control(tuner, V4L2_CID_AUDIO_VOLUME);
                return (100.0
                        * (volume - vqc->minimum)
                        / (vqc->maximum - vqc->minimum));
        } else {
                return 100.0;
        }
}

void
tuner_set_volume(struct tuner *tuner, double volume)
{
        if (tuner_has_volume_control(tuner)) {
                struct v4l2_queryctrl *vqc = &tuner->volume_ctrl;
                set_control(tuner, V4L2_CID_AUDIO_VOLUME,
                            (volume / 100.0 * (vqc->maximum - vqc->minimum)
                             + vqc->minimum));
        }
}

long long int
tuner_get_min_freq(const struct tuner *tuner)
{
        long long int rangelow = tuner->tuner.rangelow;
        if (!(tuner->tuner.capability & V4L2_TUNER_CAP_LOW))
                rangelow *= 1000;
        return rangelow;
}

long long int
tuner_get_max_freq(const struct tuner *tuner)
{
        long long int rangehigh = tuner->tuner.rangehigh;
        if (!(tuner->tuner.capability & V4L2_TUNER_CAP_LOW))
                rangehigh *= 1000;
        return rangehigh;
}

void
tuner_set_freq(const struct tuner *tuner, long long int freq,
               bool override_range)
{
        long long int adj_freq;
        struct v4l2_frequency vf;

        adj_freq = freq;
        if (!(tuner->tuner.capability & V4L2_TUNER_CAP_LOW))
                adj_freq = (adj_freq + 500) / 1000;

	if ((adj_freq < tuner->tuner.rangelow
             || adj_freq > tuner->tuner.rangehigh)
            && !override_range)
                fatal(0, "Frequency %.1f MHz out of range (%.1f - %.1f MHz)",
                      freq / 16000.0,
                      tuner_get_min_freq(tuner) / 16000.0,
                      tuner_get_max_freq(tuner) / 16000.0);

        memset(&vf, 0, sizeof vf);
        vf.tuner = tuner->index;
        vf.type = tuner->tuner.type;
        vf.frequency = adj_freq;
        if (tuner->test) {
                tuner->test->freq = adj_freq;
        }
        else if (ioctl(tuner->fd, VIDIOC_S_FREQUENCY, &vf) == -1)
                fatal(errno, "VIDIOC_S_FREQUENCY");
}

int
tuner_get_signal(const struct tuner *tuner)
{
        struct v4l2_tuner vt;

        query_tuner(tuner, &vt);
        return vt.signal;
}

void
tuner_usleep(const struct tuner *tuner, int usecs)
{
        if (!tuner->test)
                usleep(usecs);
}

void
tuner_sleep(const struct tuner *tuner, int secs)
{
        if (!tuner->test)
                sleep(secs);
}

static void
query_control(const struct tuner *tuner, uint32_t id,
              struct v4l2_queryctrl *qc)
{
        memset(qc, 0, sizeof *qc);
        qc->id = id;
        if (tuner->test) {
                assert(id == V4L2_CID_AUDIO_VOLUME);
                if (tuner->test->volume) {
                        qc->minimum = 1000;
                        qc->maximum = 2000;
                }
        } else if (ioctl(tuner->fd, VIDIOC_QUERYCTRL, qc) == -1)
                fatal(errno, "VIDIOC_QUERYCTRL");
}

static int32_t
get_control(const struct tuner *tuner, uint32_t id)
{
        struct v4l2_control control;

        memset(&control, 0, sizeof control);
        control.id = id;
        if (tuner->test) {
                assert(id == V4L2_CID_AUDIO_VOLUME);
                control.value = tuner->test->volume;
        } else if (ioctl(tuner->fd, VIDIOC_G_CTRL, &control) == -1)
                fatal(errno, "VIDIOC_G_CTRL");
        return control.value;
}

static void
set_control(const struct tuner *tuner, uint32_t id, int32_t value)
{
        struct v4l2_control control;

        memset(&control, 0, sizeof control);
        control.id = id;
        control.value = value;
        if (tuner->test) {
                if (id == V4L2_CID_AUDIO_MUTE)
                        assert(value == 0 || value == 1);
                else if (id == V4L2_CID_AUDIO_VOLUME) {
                        assert(value >= 1000 && value <= 2000);
                        tuner->test->volume = value;
                } else {
                        abort();
                }
        } else if (ioctl(tuner->fd, VIDIOC_S_CTRL, &control) == -1)
                fatal(errno, "VIDIOC_S_CTRL");
}

static void
query_tuner(const struct tuner *tuner, struct v4l2_tuner *vt)
{
        memset(vt, 0, sizeof *vt);
        vt->index = tuner->index;
        if (tuner->test) {
                int freq = tuner->test->freq;
                vt->rangelow = 16 * 89;
                vt->rangehigh = 16 * 91;
                vt->signal = (freq == (int) (16 * 89.6 + .5) ? 64000
                              : freq == (int) (16 * 90.4 + .5) ? 50000
                               : freq == (int) (16 * 90.5 + .5) ? 40000
                              : 1000);
        } else if (ioctl(tuner->fd, VIDIOC_G_TUNER, vt) == -1)
                fatal(errno, "VIDIOC_G_TUNER");
}


videodev.h
/* This header is extracted from linux/videodev.h, approximately
   version 2.6.0.  We can't use linux/videodev.h directly because
   it indirectly defines struct timespec, which is also defined
   by the standard C library headers.  Argh.  -blp */

#ifndef FM_VIDEODEV_H
#define FM_VIDEODEV_H 1

#include <stdint.h>

#define VID_TYPE_CAPTURE	1	/* Can capture */
#define VID_TYPE_TUNER		2	/* Can tune */
#define VID_TYPE_TELETEXT	4	/* Does teletext */
#define VID_TYPE_OVERLAY	8	/* Overlay onto frame buffer */
#define VID_TYPE_CHROMAKEY	16	/* Overlay by chromakey */
#define VID_TYPE_CLIPPING	32	/* Can clip */
#define VID_TYPE_FRAMERAM	64	/* Uses the frame buffer memory */
#define VID_TYPE_SCALES		128	/* Scalable */
#define VID_TYPE_MONOCHROME	256	/* Monochrome only */
#define VID_TYPE_SUBCAPTURE	512	/* Can capture subareas of the image */
#define VID_TYPE_MPEG_DECODER	1024	/* Can decode MPEG streams */
#define VID_TYPE_MPEG_ENCODER	2048	/* Can encode MPEG streams */
#define VID_TYPE_MJPEG_DECODER	4096	/* Can decode MJPEG streams */
#define VID_TYPE_MJPEG_ENCODER	8192	/* Can encode MJPEG streams */

struct video_capability
{
	char name[32];
	int type;
	int channels;	/* Num channels */
	int audios;	/* Num audio devices */
	int maxwidth;	/* Supported width */
	int maxheight;	/* And height */
	int minwidth;	/* Supported width */
	int minheight;	/* And height */
};


struct video_channel
{
	int channel;
	char name[32];
	int tuners;
	uint32_t  flags;
#define VIDEO_VC_TUNER		1	/* Channel has a tuner */
#define VIDEO_VC_AUDIO		2	/* Channel has audio */
	uint16_t  type;
#define VIDEO_TYPE_TV		1
#define VIDEO_TYPE_CAMERA	2	
	uint16_t norm;			/* Norm set by channel */
};

struct video_tuner
{
	int tuner;
	char name[32];
	unsigned long rangelow, rangehigh;	/* Tuner range */
	uint32_t flags;
#define VIDEO_TUNER_PAL		1
#define VIDEO_TUNER_NTSC	2
#define VIDEO_TUNER_SECAM	4
#define VIDEO_TUNER_LOW		8	/* Uses KHz not MHz */
#define VIDEO_TUNER_NORM	16	/* Tuner can set norm */
#define VIDEO_TUNER_STEREO_ON	128	/* Tuner is seeing stereo */
#define VIDEO_TUNER_RDS_ON      256     /* Tuner is seeing an RDS datastream */
#define VIDEO_TUNER_MBS_ON      512     /* Tuner is seeing an MBS datastream */
	uint16_t mode;			/* PAL/NTSC/SECAM/OTHER */
#define VIDEO_MODE_PAL		0
#define VIDEO_MODE_NTSC		1
#define VIDEO_MODE_SECAM	2
#define VIDEO_MODE_AUTO		3
	uint16_t signal;			/* Signal strength 16bit scale */
};

struct video_picture
{
	uint16_t	brightness;
	uint16_t	hue;
	uint16_t	colour;
	uint16_t	contrast;
	uint16_t	whiteness;	/* Black and white only */
	uint16_t	depth;		/* Capture depth */
	uint16_t   palette;	/* Palette in use */
#define VIDEO_PALETTE_GREY	1	/* Linear greyscale */
#define VIDEO_PALETTE_HI240	2	/* High 240 cube (BT848) */
#define VIDEO_PALETTE_RGB565	3	/* 565 16 bit RGB */
#define VIDEO_PALETTE_RGB24	4	/* 24bit RGB */
#define VIDEO_PALETTE_RGB32	5	/* 32bit RGB */	
#define VIDEO_PALETTE_RGB555	6	/* 555 15bit RGB */
#define VIDEO_PALETTE_YUV422	7	/* YUV422 capture */
#define VIDEO_PALETTE_YUYV	8
#define VIDEO_PALETTE_UYVY	9	/* The great thing about standards is ... */
#define VIDEO_PALETTE_YUV420	10
#define VIDEO_PALETTE_YUV411	11	/* YUV411 capture */
#define VIDEO_PALETTE_RAW	12	/* RAW capture (BT848) */
#define VIDEO_PALETTE_YUV422P	13	/* YUV 4:2:2 Planar */
#define VIDEO_PALETTE_YUV411P	14	/* YUV 4:1:1 Planar */
#define VIDEO_PALETTE_YUV420P	15	/* YUV 4:2:0 Planar */
#define VIDEO_PALETTE_YUV410P	16	/* YUV 4:1:0 Planar */
#define VIDEO_PALETTE_PLANAR	13	/* start of planar entries */
#define VIDEO_PALETTE_COMPONENT 7	/* start of component entries */
};

struct video_audio
{
	int	audio;		/* Audio channel */
	uint16_t	volume;		/* If settable */
	uint16_t	bass, treble;
	uint32_t	flags;
#define VIDEO_AUDIO_MUTE	1
#define VIDEO_AUDIO_MUTABLE	2
#define VIDEO_AUDIO_VOLUME	4
#define VIDEO_AUDIO_BASS	8
#define VIDEO_AUDIO_TREBLE	16	
#define VIDEO_AUDIO_BALANCE	32
	char    name[16];
#define VIDEO_SOUND_MONO	1
#define VIDEO_SOUND_STEREO	2
#define VIDEO_SOUND_LANG1	4
#define VIDEO_SOUND_LANG2	8
        uint16_t   mode;
        uint16_t	balance;	/* Stereo balance */
        uint16_t	step;		/* Step actual volume uses */
};

struct video_clip
{
	int32_t	x,y;
	int32_t	width, height;
	struct	video_clip *next;	/* For user use/driver use only */
};

struct video_window
{
	uint32_t	x,y;			/* Position of window */
	uint32_t	width,height;		/* Its size */
	uint32_t	chromakey;
	uint32_t	flags;
	struct	video_clip *clips;	/* Set only */
	int	clipcount;
#define VIDEO_WINDOW_INTERLACE	1
#define VIDEO_WINDOW_CHROMAKEY	16	/* Overlay by chromakey */
#define VIDEO_CLIP_BITMAP	-1
/* bitmap is 1024x625, a '1' bit represents a clipped pixel */
#define VIDEO_CLIPMAP_SIZE	(128 * 625)
};

struct video_capture
{
	uint32_t 	x,y;			/* Offsets into image */
	uint32_t	width, height;		/* Area to capture */
	uint16_t	decimation;		/* Decimation divider */
	uint16_t	flags;			/* Flags for capture */
#define VIDEO_CAPTURE_ODD		0	/* Temporal */
#define VIDEO_CAPTURE_EVEN		1
};

struct video_buffer
{
	void	*base;
	int	height,width;
	int	depth;
	int	bytesperline;
};

struct video_mmap
{
	unsigned	int frame;		/* Frame (0 - n) for double buffer */
	int		height,width;
	unsigned	int format;		/* should be VIDEO_PALETTE_* */
};

struct video_key
{
	uint8_t	key[8];
	uint32_t	flags;
};


#define VIDEO_MAX_FRAME		32

struct video_mbuf
{
	int	size;		/* Total memory to map */
	int	frames;		/* Frames */
	int	offsets[VIDEO_MAX_FRAME];
};
	

#define 	VIDEO_NO_UNIT	(-1)

	
struct video_unit
{
	int 	video;		/* Video minor */
	int	vbi;		/* VBI minor */
	int	radio;		/* Radio minor */
	int	audio;		/* Audio minor */
	int	teletext;	/* Teletext minor */
};

struct vbi_format {
	uint32_t	sampling_rate;	/* in Hz */
	uint32_t	samples_per_line;
	uint32_t	sample_format;	/* VIDEO_PALETTE_RAW only (1 byte) */
	int32_t	start[2];	/* starting line for each frame */
	uint32_t	count[2];	/* count of lines for each frame */
	uint32_t	flags;
#define	VBI_UNSYNC	1	/* can distingues between top/bottom field */
#define	VBI_INTERLACED	2	/* lines are interlaced */
};

/* video_info is biased towards hardware mpeg encode/decode */
/* but it could apply generically to any hardware compressor/decompressor */
struct video_info
{
	uint32_t	frame_count;	/* frames output since decode/encode began */
	uint32_t	h_size;		/* current unscaled horizontal size */
	uint32_t	v_size;		/* current unscaled veritcal size */
	uint32_t	smpte_timecode;	/* current SMPTE timecode (for current GOP) */
	uint32_t	picture_type;	/* current picture type */
	uint32_t	temporal_reference;	/* current temporal reference */
	uint8_t	user_data[256];	/* user data last found in compressed stream */
	/* user_data[0] contains user data flags, user_data[1] has count */
};

/* generic structure for setting playback modes */
struct video_play_mode
{
	int	mode;
	int	p1;
	int	p2;
};

/* for loading microcode / fpga programming */
struct video_code
{
	char	loadwhat[16];	/* name or tag of file being passed */
	int	datasize;
	uint8_t	*data;
};

#define VIDIOCGCAP		_IOR('v',1,struct video_capability)	/* Get capabilities */
#define VIDIOCGCHAN		_IOWR('v',2,struct video_channel)	/* Get channel info (sources) */
#define VIDIOCSCHAN		_IOW('v',3,struct video_channel)	/* Set channel 	*/
#define VIDIOCGTUNER		_IOWR('v',4,struct video_tuner)		/* Get tuner abilities */
#define VIDIOCSTUNER		_IOW('v',5,struct video_tuner)		/* Tune the tuner for the current channel */
#define VIDIOCGPICT		_IOR('v',6,struct video_picture)	/* Get picture properties */
#define VIDIOCSPICT		_IOW('v',7,struct video_picture)	/* Set picture properties */
#define VIDIOCCAPTURE		_IOW('v',8,int)				/* Start, end capture */
#define VIDIOCGWIN		_IOR('v',9, struct video_window)	/* Get the video overlay window */
#define VIDIOCSWIN		_IOW('v',10, struct video_window)	/* Set the video overlay window - passes clip list for hardware smarts , chromakey etc */
#define VIDIOCGFBUF		_IOR('v',11, struct video_buffer)	/* Get frame buffer */
#define VIDIOCSFBUF		_IOW('v',12, struct video_buffer)	/* Set frame buffer - root only */
#define VIDIOCKEY		_IOR('v',13, struct video_key)		/* Video key event - to dev 255 is to all - cuts capture on all DMA windows with this key (0xFFFFFFFF == all) */
#define VIDIOCGFREQ		_IOR('v',14, unsigned long)		/* Set tuner */
#define VIDIOCSFREQ		_IOW('v',15, unsigned long)		/* Set tuner */
#define VIDIOCGAUDIO		_IOR('v',16, struct video_audio)	/* Get audio info */
#define VIDIOCSAUDIO		_IOW('v',17, struct video_audio)	/* Audio source, mute etc */
#define VIDIOCSYNC		_IOW('v',18, int)			/* Sync with mmap grabbing */
#define VIDIOCMCAPTURE		_IOW('v',19, struct video_mmap)		/* Grab frames */
#define VIDIOCGMBUF		_IOR('v',20, struct video_mbuf)		/* Memory map buffer info */
#define VIDIOCGUNIT		_IOR('v',21, struct video_unit)		/* Get attached units */
#define VIDIOCGCAPTURE		_IOR('v',22, struct video_capture)	/* Get subcapture */
#define VIDIOCSCAPTURE		_IOW('v',23, struct video_capture)	/* Set subcapture */
#define VIDIOCSPLAYMODE		_IOW('v',24, struct video_play_mode)	/* Set output video mode/feature */
#define VIDIOCSWRITEMODE	_IOW('v',25, int)			/* Set write mode */
#define VIDIOCGPLAYINFO		_IOR('v',26, struct video_info)		/* Get current playback info from hardware */
#define VIDIOCSMICROCODE	_IOW('v',27, struct video_code)		/* Load microcode into hardware */
#define	VIDIOCGVBIFMT		_IOR('v',28, struct vbi_format)		/* Get VBI information */
#define	VIDIOCSVBIFMT		_IOW('v',29, struct vbi_format)		/* Set VBI information */


#define BASE_VIDIOCPRIVATE	192		/* 192-255 are private */

/* VIDIOCSWRITEMODE */
#define VID_WRITE_MPEG_AUD		0
#define VID_WRITE_MPEG_VID		1
#define VID_WRITE_OSD			2
#define VID_WRITE_TTX			3
#define VID_WRITE_CC			4
#define VID_WRITE_MJPEG			5

/* VIDIOCSPLAYMODE */
#define VID_PLAY_VID_OUT_MODE		0
	/* p1: = VIDEO_MODE_PAL, VIDEO_MODE_NTSC, etc ... */
#define VID_PLAY_GENLOCK		1
	/* p1: 0 = OFF, 1 = ON */
	/* p2: GENLOCK FINE DELAY value */ 
#define VID_PLAY_NORMAL			2
#define VID_PLAY_PAUSE			3
#define VID_PLAY_SINGLE_FRAME		4
#define VID_PLAY_FAST_FORWARD		5
#define VID_PLAY_SLOW_MOTION		6
#define VID_PLAY_IMMEDIATE_NORMAL	7
#define VID_PLAY_SWITCH_CHANNELS	8
#define VID_PLAY_FREEZE_FRAME		9
#define VID_PLAY_STILL_MODE		10
#define VID_PLAY_MASTER_MODE		11
	/* p1: see below */
#define		VID_PLAY_MASTER_NONE	1
#define		VID_PLAY_MASTER_VIDEO	2
#define		VID_PLAY_MASTER_AUDIO	3
#define VID_PLAY_ACTIVE_SCANLINES	12
	/* p1 = first active; p2 = last active */
#define VID_PLAY_RESET			13
#define VID_PLAY_END_MARK		14



#define VID_HARDWARE_BT848	1
#define VID_HARDWARE_QCAM_BW	2
#define VID_HARDWARE_PMS	3
#define VID_HARDWARE_QCAM_C	4
#define VID_HARDWARE_PSEUDO	5
#define VID_HARDWARE_SAA5249	6
#define VID_HARDWARE_AZTECH	7
#define VID_HARDWARE_SF16MI	8
#define VID_HARDWARE_RTRACK	9
#define VID_HARDWARE_ZOLTRIX	10
#define VID_HARDWARE_SAA7146    11
#define VID_HARDWARE_VIDEUM	12	/* Reserved for Winnov videum */
#define VID_HARDWARE_RTRACK2	13
#define VID_HARDWARE_PERMEDIA2	14	/* Reserved for Permedia2 */
#define VID_HARDWARE_RIVA128	15	/* Reserved for RIVA 128 */
#define VID_HARDWARE_PLANB	16	/* PowerMac motherboard video-in */
#define VID_HARDWARE_BROADWAY	17	/* Broadway project */
#define VID_HARDWARE_GEMTEK	18
#define VID_HARDWARE_TYPHOON	19
#define VID_HARDWARE_VINO	20	/* SGI Indy Vino */
#define VID_HARDWARE_CADET	21	/* Cadet radio */
#define VID_HARDWARE_TRUST	22	/* Trust FM Radio */
#define VID_HARDWARE_TERRATEC	23	/* TerraTec ActiveRadio */
#define VID_HARDWARE_CPIA	24
#define VID_HARDWARE_ZR36120	25	/* Zoran ZR36120/ZR36125 */
#define VID_HARDWARE_ZR36067	26	/* Zoran ZR36067/36060 */
#define VID_HARDWARE_OV511	27	
#define VID_HARDWARE_ZR356700	28	/* Zoran 36700 series */
#define VID_HARDWARE_W9966	29
#define VID_HARDWARE_SE401	30	/* SE401 USB webcams */
#define VID_HARDWARE_PWC	31	/* Philips webcams */
#define VID_HARDWARE_MEYE	32	/* Sony Vaio MotionEye cameras */
#define VID_HARDWARE_CPIA2	33
#define VID_HARDWARE_VICAM      34
#define VID_HARDWARE_SF16FMR2	35

#endif /* videodev.h */
